### DISCUSSION TITLE
An example or two for configuring rust-analyzer?

### DISCUSSION THREAD

**Discussion by jkaye2012** [mrcjkb/rustaceanvim#319] [Q&A] [answered]
### Feature description

Hi,

I understand that features such as autoimport on completion are rust-analyzer features rather than something provided by this plugin; however, I'm finding it _very_ difficult to understand how one is actually meant to set such configurations given the 3-4 different pieces of documentation that other issues point to. I haven't encountered any example of a realistic configuration that would be useful as a starting point.

For example, I've been trying a few different iterations of this with no success whatsoever (no error messages, but the functionality is still not working):

```
vim.g.rustaceanvim = {
  -- Plugin configuration
  tools = {
  },
  -- LSP configuration
  server = {
    on_attach = function(client, bufnr)
      -- you can also put keymaps in here
    end,
    default_settings = {
      ['rust-analyzer'] = {
        completion = {
          autoimport = {
            enable = true,
          }
        },
      },
    },
  },
  -- DAP configuration
  dap = {
  },
}
```

If there are such examples somewhere and I've simply missed them, please point me to them!

Thanks.

**Comment by mrcjkb** [comment, marked as answer]
Hey 👋 

You shouldn't need to configure rust-analyzer to enable that.
But import on completion isn't natively supported by Neovim. You need a completion plugin for that.
For example, [nvim-cmp](https://github.com/hrsh7th/nvim-cmp) with the [cmp-nvim-lsp extension](https://github.com/hrsh7th/cmp-nvim-lsp).
And you need to make sure you register the plugin's capabilities with the `server.capabilities` field.

**Reply by jkaye2012** [reply]
This was helpful, thank you.

On an only partially-related note, are you aware of anywhere that this type of information is documented in general for someone with a good amount of editor configuration experience, but no Neovim experience? I've tried reading multiple manuals, but at least so far as I can tell, this type of information is "locked away" in tacit knowledge of individuals who are familiar with how the ecosystem is meant to work together. Many of the issues I've seen in this repository and others like it seem to indicate that I'm not alone in this experience. This certainly isn't a problem with rustaceanvim - more of an issue with Neovim itself and the current state of the tooling - but I was thinking you might know the answer to some of these questions!

It's been a frustrating experience for me, so I'm considering doing a good amount of research so that I can create some better documentation for the current state of the ecosystem, but if such a thing already exists I don't want to duplicate the work.

**Reply by mrcjkb** [reply]
Here are some up-to-date (as of writing this comment) resources I know of:

- @vhyrro has made [a great series on YouTube](https://www.youtube.com/watch?v=87AXw9Quy9U&list=PLx2ksyallYzW4WNYHD9xOFrPRYGlntAft&pp=iAQB), which includes [a video that goes in depth into Neovim's LSP client](https://youtu.be/HL7b63Hrc8U?si=0n0ORDiAw5KoGivr).

- [kickstart.nvim](https://github.com/nvim-lua/kickstart.nvim) is a very popular starter config with very detailed comments explaining everything. 
  Although I personally don't agree with some of its choices, i'd recommend it for newcomers over any pre-configured Neovim distro I'm aware of.
  
- It's definitely a good idea to get comfortable navigating Neovim's help pages (e.g. , you could type `:h completion` or `:h autocomplete` and `:h lsp` to see what Neovim has to offer without any plugins).
  For plugins, you can type `:h <plugin-name>` (e.g. `:h rustaceanvim`).
  Tip: In the help pages, you can use `<C-]>` to jump to the help page for the word your cursor is on, and `<C-t>` to jump back.

Apart from that, it just takes time to figure everything out. IMO, it's okay to ask if you can't figure it out yourself. In the worst case, someone might get annoyed, but they'll probably still tell you where to look :smile:

**Reply by Syphixs** [reply]
Hey there I have the same struggles as you. Did you manage to get a working useful example config which you are maybe able to share? I am quite new to nvim but the nvim-cmp setting to work with this plugin is still a mystery to me.  If you cannot share your config can you maybe provide me some useful hint to a documentation please. I looked at the videos but they seem very advanced sadly

Update: Found a very helpful link for all still struggling:
https://www.lazyvim.org/extras/lang/rust

**Reply by jkaye2012** [reply]
I did, at least in my setup! Here is my [completion config](https://github.com/jkaye2012/neovim-flake/blob/main/config/completion.lua) and [lsp config](https://github.com/jkaye2012/neovim-flake/blob/main/config/lsp.lua). I don't have anything else configured (other than having `rust-analyzer` and friends installed) and everything is working for me as I'd expect. I still don't love the keybindings and there are other changes that I'll be making in the future as well, feel free to follow the repo if you'd like.
